File: 0019-Remove-Nth-Node-From-End-of-List/lc0019.cc
// C++ solution.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode dumb(0);
        dumb.next = head;
        ListNode* p1 = &dumb;
        ListNode* p2 = &dumb;
        while ((n--) >= 0) p2 = p2 -> next;
        while (p2 != nullptr) {
            p1 = p1 -> next;
            p2 = p2 -> next;
        }
        p1 -> next = p1 -> next -> next;
        return dumb.next;
    }
};

File: 0019-Remove-Nth-Node-From-End-of-List/lc0019.rs
// Rust solution.

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut dumb = Some(Box::new(ListNode {val: 0, next: head}));
        let mut length = 0;
        {
            let mut p = dumb.as_ref();
            while p.unwrap().next.is_some() {
                p = p.unwrap().next.as_ref();
                length += 1;
            }
        }
        let mut p = dumb.as_mut();
        for _ in 0..(length-n) {
            p = p.unwrap().next.as_mut();
        }
        p.unwrap().next = p.as_mut().unwrap().next.as_mut().unwrap().next.take();
        dumb.unwrap().next
    }
}
